firebase-firestore-mixin.html

```diff
@@ -0,0 +1,315 @@
+   *   the element instance. It must return an instance of Query.
+   * * **observes:** a list of properties which, if changed, should trigger
+   *   a rebuild of a listener.
+   * * **noCache:** cached Firestore data won't be assigned to a property
+   *   value even if persistence is enabled.
+   *
+   * @polymer
+   * @mixinFunction Polymer.FirestoreMixin
+   */
+  Polymer.FirestoreMixin = parent =>
+    class extends parent {
+      static _assertPropertyTypeCorrectness(prop) {
+        const errorMessage = (listenerType, propertyType) =>
+          `FirestoreMixin's ${listenerType} can only be used with properties ` +
+          `of type ${propertyType}.`;
+        const assert = (listenerType, propertyType) => {
+          if (prop[listenerType] !== undefined && prop.type !== propertyType) {
+            throw new Error(errorMessage(listenerType, propertyType.name));
+          }
+        }
+
+        assert('doc', Object);
+        assert('collection', Array);
+      }
+
+      constructor() {
+        super();
+
+        if (this[CONSTRUCTOR_TOKEN] === true) {
+          return;
+        }
+        this[CONSTRUCTOR_TOKEN] = true;
+
+        this._firestoreProps = {};
+        this._firestoreListeners = {};
+        this.db = this.constructor.db || firebase.firestore();
+      }
+
+      connectedCallback() {
+        if (this[CONNECTED_CALLBACK_TOKEN] !== true) {
+          this[CONNECTED_CALLBACK_TOKEN] = true;
+
+          const props = collect(this.constructor, 'properties');
+          Object
+            .values(props)
+            .forEach(this.constructor._assertPropertyTypeCorrectness);
+
+          for (let name in props) {
+            const options = props[name];
+            if (options.doc || options.collection) {
+              this._firestoreBind(name, options);
+            }
+          }
+        }
+
+        super.connectedCallback();
+      }
+
+      _firestoreBind(name, options) {
+        const defaults = {
+          live: false,
+          observes: [],
+        }
+        const parsedPath = parsePath(options.doc || options.collection);
+        const config = Object.assign({}, defaults, options, parsedPath);
+        const type = config.type =
+          config.doc ? 'doc' : config.collection ? 'collection' : undefined;
+
+        this._firestoreProps[name] = config;
+
+        const args = config.props.concat(config.observes);
+        if (args.length > 0) {
+          // Create a method observer that will be called every time
+          // a templatized or observed property changes
+          const observer =
+            `_firestoreUpdateBinding('${name}', ${args.join(',')})`
+          this._createMethodObserver(observer);
+        }
+
+        this._firestoreUpdateBinding(name, ...args.map(x => this[x]));
+      }
+
+      _firestoreUpdateBinding(name, ...args) {
+        this._firestoreUnlisten(name);
+
+        const config = this._firestoreProps[name];
+        const isDefined = (x) => x !== undefined;
```

Comment:
Question here around whether the mixin should defensively prevent the binding of the DB when `propArgs[n] === ''` as well as when `propArgs[n] === undefined`? Theoretically it can be programmed around, and it publishes a clear error up the line, but it seems like here's a pretty solid place to prevent complaints, as well as make possible some potentially useful patterns.